File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ser/binary/types/VertexPropertySerializer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.driver.ser.binary.types;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufAllocator;
import io.netty.buffer.CompositeByteBuf;
import org.apache.tinkerpop.gremlin.driver.ser.SerializationException;
import org.apache.tinkerpop.gremlin.driver.ser.binary.DataType;
import org.apache.tinkerpop.gremlin.driver.ser.binary.GraphBinaryReader;
import org.apache.tinkerpop.gremlin.driver.ser.binary.GraphBinaryWriter;
import org.apache.tinkerpop.gremlin.structure.VertexProperty;
import org.apache.tinkerpop.gremlin.structure.util.reference.ReferenceVertexProperty;

/**
 * @author Stephen Mallette (http://stephen.genoprime.com)
 */
public class VertexPropertySerializer extends SimpleTypeSerializer<VertexProperty> {

    public VertexPropertySerializer() {
        super(DataType.VERTEXPROPERTY);
    }

    @Override
    VertexProperty readValue(final ByteBuf buffer, final GraphBinaryReader context) throws SerializationException {
        final VertexProperty v = new ReferenceVertexProperty<>(context.read(buffer),
                context.readValue(buffer, String.class, false),
                context.read(buffer));

        // discard the parent vertex - we only send "references"
        context.read(buffer);

        // discard the properties - as we only send "references" this should always be null, but will we change our
        // minds some day????
        context.read(buffer);
        return v;
    }

    @Override
    public ByteBuf writeValue(final VertexProperty value, final ByteBufAllocator allocator, final GraphBinaryWriter context) throws SerializationException {
        final CompositeByteBuf result = allocator.compositeBuffer(5);
        result.addComponent(true, context.write(value.id(), allocator));
        result.addComponent(true, context.writeValue(value.label(), allocator, false));
        result.addComponent(true, context.write(value.value(), allocator));

        // we don't serialize the parent vertex even as a "reference", but, let's hold a place for it
        result.addComponent(true, context.write(null, allocator));
        // we don't serialize properties for graph elements. they are "references", but we leave a place holder
        // here as an option for the future as we've waffled this soooooooooo many times now
        result.addComponent(true, context.write(null, allocator));

        return result;
    }
}

File: gremlin-js/gremlin-javascript/lib/structure/io/binary/internals/VertexPropertySerializer.js
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

/**
 * @author Igor Ostapenko
 */

import { Buffer } from 'buffer';
import { VertexProperty } from '../../../graph.js';

export default class VertexPropertySerializer {
  constructor(ioc) {
    this.ioc = ioc;
    this.ioc.serializers[ioc.DataType.VERTEXPROPERTY] = this;
  }

  canBeUsedFor(value) {
    return value instanceof VertexProperty;
  }

  serialize(item, fullyQualifiedFormat = true) {
    if (item === undefined || item === null) {
      if (fullyQualifiedFormat) {
        return Buffer.from([this.ioc.DataType.VERTEXPROPERTY, 0x01]);
      }
      return Buffer.concat([
        this.ioc.unspecifiedNullSerializer.serialize(null), // {id}=null
        this.ioc.listSerializer.serialize([], false), // {label}=empty list
        this.ioc.unspecifiedNullSerializer.serialize(null), // {value}=null
        this.ioc.unspecifiedNullSerializer.serialize(null), // {parent}=null
        this.ioc.listSerializer.serialize([], true), // {properties}=empty list
      ]);
    }

    const bufs = [];
    if (fullyQualifiedFormat) {
      bufs.push(Buffer.from([this.ioc.DataType.VERTEXPROPERTY, 0x00]));
    }

    // {id}
    bufs.push(this.ioc.anySerializer.serialize(item.id));

    // {label}
    const labels = Array.isArray(item.label) ? item.label : item.label ? [item.label] : [];
    bufs.push(this.ioc.listSerializer.serialize(labels, false));

    // {value}
    bufs.push(this.ioc.anySerializer.serialize(item.value));

    // {parent}
    bufs.push(this.ioc.unspecifiedNullSerializer.serialize(null));

    // {properties}
    bufs.push(this.ioc.listSerializer.serialize([], true));

    return Buffer.concat(bufs);
  }

  /**
   * Async deserialization of vertex property value bytes from a StreamReader.
   * @param {StreamReader} reader
   * @param {number} valueFlag
   * @param {number} typeCode
   * @returns {Promise<VertexProperty>}
   */
  async deserializeValue(reader, valueFlag, typeCode) {
    // {id} fully qualified
    const id = await this.ioc.anySerializer.deserialize(reader);

    // {label} bare list, extract first element
    const labelList = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);
    const label = Array.isArray(labelList) && labelList.length > 0 ? labelList[0] : labelList;

    // {value} fully qualified
    const value = await this.ioc.anySerializer.deserialize(reader);

    // {parent} fully qualified (always null in current TinkerPop)
    await this.ioc.anySerializer.deserialize(reader);

    // {properties} fully qualified
    const properties = await this.ioc.anySerializer.deserialize(reader);

    return new VertexProperty(id, label, value, properties || []);
  }

  /**
   * Async fully-qualified deserialization from a StreamReader.
   * @param {StreamReader} reader
   * @returns {Promise<VertexProperty|null>}
   */
  async deserialize(reader) {
    const type_code = await reader.readUInt8();
    if (type_code !== this.ioc.DataType.VERTEXPROPERTY) {
      throw new Error(`VertexPropertySerializer: unexpected {type_code}=0x${type_code.toString(16)}`);
    }
    const value_flag = await reader.readUInt8();
    if (value_flag === 0x01) {
      return null;
    }
    if (value_flag !== 0x00) {
      throw new Error(`VertexPropertySerializer: unexpected {value_flag}=0x${value_flag.toString(16)}`);
    }
    return this.deserializeValue(reader, value_flag, type_code);
  }
}
